### DISCUSSION TITLE
Late arriving data - by design - an IoT use case

### DISCUSSION THREAD

**Discussion by Recodify** [questdb/questdb#2503] [Show and tell]
Hi everyone,

Not really a show and tell, but more a show and ask. I toyed with putting this discussion in slack, but I think what I'm going to ask might be useful to others and things tend to get buried in slack.

### My Use Case.

- IoT
- Temperature and humidity sensors deployed in residential properties.
- Sensors communicate to an in-home hub via RF 868.
- In-home hub has a GSM backhaul/connection to the internet.
- Sensors take a reading every 15mins and communicate this reading to the in-home hub instantly. 
- In-home hub caches the readings for 2 hours to reduce data costs over the cellular network due to overhead of communication i.e. fewer transmits with larger payloads (more readings) is more data efficient.
- The 2 hour delay is not synchronised i.e. an in-home hub can and does send it's data every second of every day (we have a lot of hubs!)

So, in summary, delivery of data across the network is "laggy" by design. This lag is obviously _significantly_ longer than the type of lag you would expect to see with typical TCP based delivery due to network jitter and out-of-order (OOO) async processing. 

### My Use Case - In QuestDb

I see a few possible approaches.

#### Option 1 - long commit lag

Set a commit lag in QuestDb to cover the 2 hour window + a little extra to account for the standard jitter once the message arrives on a TCP network.

This feels like the best approach for optimal data ingest but feels like it might have some potential pain points.

1. My server would need enough RAM to store a two hour block of readings + overhead for normal operation. Whilst this _might_ be possible today, it doesn't feel very scalable as the number of sensors grows overtime.
2. Risk of data loss - the more data we have in-memory and uncommitted the greater the chance we might lose that data in the event of an unexpected shutdown - power loss etc.

#### Option 2 - slower ingest

Accept that all my ingests will incur the penalty of data file resorts

1. Feels safer but not very satisfying. We're taking something which is amazingly good (the super fast ingest) and throwing it away. I guess the importance of this depends entirely on just how big a penalty we pay from the resorts.
2. Do these resorts result in degraded read performance?

#### Option 3 - I'm missing something

I hope this is the case! Is there a third option which I've not considered?

Would love to get everyones thoughts.

**Comment by javier** [comment]
This will probably be a non-issue in the near future. We are working on the WAL (write ahead log) functionality to allow for read replicas. The way WAL works, all the data we receive is first stored in the WAL files, then committed to the TableWriter. This means that non committed data will still be stored in the WAL, and also means when TableWriter is committing data, ingestion is decoupled, so less affected by potential out of order problems.

WAL is (hopefully) not even a few months away. For the time being, I don't see any options other than the ones you described.
